tools/rapidrun.1.md
% RAPIDRUN(1) Rapicorn-@BUILDID@ | Rapicorn Manual
%
% @FILE_REVISION@


# NAME
rapidrun - Utility to run and test Rapicorn dialogs


# SYNOPSIS
**rapidrun** [*OPTIONS*] [*uifile.xml*]


# DESCRIPTION

**Rapidrun** is a utility to display and operate Rapicorn UI dialogs.
This tool will read the GUI description file listed on the command line, look for a dialog window and show it.


# OPTIONS

**Rapidrun** follows the usual GNU command line syntax, with long options starting with two dashes ('-').

**\--parse-test**
:   Parse GuiFile.xml and exit.

**-x**
:   Enable auto-exit after first expose.

**\--list**
:   List parsed definitions.

**\--fatal-warnings**
:   Turn criticals/warnings into fatal conditions.

**\--snapshot** *pngname*
:   Dump a snapshot to *pngname*.

**\--test-dump**
:   Dump test stream after first expose.

**\--test-matched-node** *PATTERN*
:   Filter nodes in test dumps.

**-h**, **\--help**
:   Display this help and exit.

**-v**, **\--version**
:   Display version and exit.


# ENVIRONMENT VARIABLES

The behavior of **rapidrun** is affected by the following environment variables.

**RAPICORN_DEBUG**
:   This environment variable affects debugging behavior for all Rapicorn applications.
    Some possible values are:
    *all* to enable all available debugging output,
    *syslog* to enable logging of general purpose messages through syslog(3),
    *fatal-warnings* to cast all warning messages into fatal errors,
    *devel* to enable some debugging features for development versions.
    Refer to the Rapicorn Manual for a more detailed description.

**RAPIDRUN_RES**
:   This environment variable contains a search path used by unit tests for test images
    residing in the file system instead of compiled in resources.


# AUTHORS
Tim Janik.


# SEE ALSO

**rapidres**(1), [**Rapicorn Website**](https://rapicorn.testbit.org)


NEWS
Rapicorn 17.0.0~rc1:

* Support clang++-3.6.2.
* Fixed uild on Arch linux [milkmiruku]
* Fixed build on Kubuntu 16.04 [stw]
* Fixed bad_weak_ptr exceptions.
* Added support g_main_context() integration.
* Improved the quality of rendering stretched SVG files.
* Fixed spacing calculcations, text block creation and resizing priorities.
* Parse .ini files from $XDG_CONFIG_HOME, $XDG_CONFIG_DIRS and /etc/rapicorn/.
* Support selectable container children.
* Support Capturing event phase.
* Automatically allow shrinking for all spreading widgets to preserve others.
* Fixed focus movemenet to (partially) obscured widgets.
* Use viewport window relative coordinates for all widgets.
* Simplified code in several places.
* Improved rendering speed significantly.
* Introduced SelectableItem for lists and remove some dated code.
* Cleaned up the ancestry cache handling and allow Window.destroy().
* Simplified ThemeInfo and styling.
* Added support for SVGs in the widget factory.
* Use 16 bit for widget states and add TOGGLED and STASHED.
* Fix the SVG ink rectangle where libsvg only exports ints instead of doubles.
* Add ElementPainter: with 'element' and 'filler' properties to render SVGs.
* Allow normal buttons to toggle.
* Added ShapePainter that can render circle, diamond, square and star.
* Support a Button.label property.
* Add a RadioButton implementation.
* Improved filewview UI.
* SVG theme updates.
* Use <Element id="some-identifier" declare="new-element"/> in XML UIs.
* Added Widget properties acceleratable, hover, panel, receives_default.
* Also add Widget properties selected, insensitive and active.
* Consistently use int for size calculations everywhere.
* Optimized resizing and invalidation region calculations.
* Cleaned up the code and build rules in several places.
* Support fine grained size negotiations.
* Support per widget offline surfaces.
* Better documented ABI versioning affects.
* Improved documentation.
* Added test categorization and numbering.
* Build Rapicorn via non-recursive sub Makefiles.
* Fixed use-after-free in DisplayWindowX11.
* Avoid overlinking in rapicorn.pc.
* Require srcdir=destdir upon distcheck.
* Added backtrace generator via addr2line.
* Added support for notifications from Aida remote calls.
* Added support for event loop source groups.
* Majorly improved Docker build setup.

Rapicorn 16.0.0:	(2016-04-07)

* Remove internal librsvg copy, require an installed librsvg-2.0.
* Provide enum meta information for enum values stored in Any.
* Provide change notification signals for Parameter.
* Cleanup threads via atexit(3).
* Added support for auxillary data on IDL sequence and record definitions.
* Added compatibility code to downgrade cython requirement to 0.20.1 for trusty.
* Added fixes for g++-4.8.4 and libtool >= 1.5.2.
* Grow Any content to enable structured type and small string optimizations.
* Remove remove Atomic<> in favour of the "Construct On First Use Idiom".
* Improved exception specifications to satisfy cython-0.23.3.
* Reduced unit testing time, individually and by tuning the benchmark timers.
* Use xvfb-run(1) to run X11 related unit tests.
* Support and use the TAP protocol for unit test evaluation.
* New versioning: Releases increase MAJOR for major API changes.
* New versioning: Releases increase MINOR in case of any ABI breaks.
* New versioning: Releases increase MICRO for backwards compatibility.
* New versioning: The MAJOR number ideally matches the release year.
* The Python module alias remans 'Rapicorn', stored in Rapicorn_15/.
* C++ namespace changed to 'Rapicorn', files install under 'rapicorn-$MAJOR'.
* The librapicorn shared library supports installation of multiple versions.
* Many build system fixes and integration of travis-ci builds.
* Automatically built Debian packages are now on http://bintray.com/beast-team/.
* The file debian/copyright now contains an accurate licensing account.
* Install cython headers to allow IDL Python bindings outside Rapicorn.
* Add AutoSeeder class to provide good non-deterministic seeding entropy.
* Add PCG-32 which provides fast but well distributed pseudo random numbers.
* Include a random_secret() in hash functions to avoid collision attacks.
* Use Python enum classes and C++11 enum classes in IDL code generation.
* Fixed shebang to reliably find Python2.7. [Milk Brewster]

Rapicorn 15.09.0:	(2015-09-16)

* Added theming support and $RAPICORN_THEME to override theme selection.
* Update RSVG rendering bits.
* Added layer and state painting widgets.
* Provide rapidres(1): A binary resource file to C++ code converter.
* Added missing manual page documentation.
* Added ability to render SVG fragments.
* Added widget for state painting and for stretched SVG painting.
* Added ElementPainter to render and resize widgets from stretched SVGs.
* Added FocusPainter for focus state dependent decoration.
* Added support for attribute="@eval some_variable" syntax.
* Added example for interactive code editing and execution.
* Factory cleanups, widget have an 'id', properties have a 'name' attribute.
* Allow source priorities to affect scheduling across other event loops.
* Major improvements to the Aida Python bindings by making use of Cython.
* IDL records and sequences now have different client and server code.
* Eliminated '=0' syntax for IDL methods.
* Added C++ visitors to IDL objects, records and sequences for INI and XML.
* Improved enum introspection and IDL marshalling.
* Major improvements to type support in Any getters and setters.
* Provide forward and backward conversion between Any and records/sequences.
* Added Parameter abstraction to simlify record field and property edits.
* Display Window code cleanup and optimization.
* Automate memory bookkeeping with shared_ptr for all objects and widgets.
* Various improvements to the C++11 support, compilaion and build system.

Rapicorn 14.10.0:	(2014-10-09)

* Project is re-license under MPLv2.

tools/rapidrun.cc
// This Source Code Form is licensed MPL-2.0: http://mozilla.org/MPL/2.0
#include <rapicorn.hh>

#include <string.h>
#include <stdlib.h>

namespace {
using namespace Rapicorn;

static void
help_usage (bool usage_error)
{
  const char *usage = "Usage: rapidrun [OPTIONS] <GuiFile.xml>";
  if (usage_error)
    {
      printerr ("%s\n", usage);
      printerr ("Try 'rapidrun --help' for more information.\n");
      exit (1);
    }
  printout ("%s\n", usage);
  /*         12345678901234567890123456789012345678901234567890123456789012345678901234567890 */
  printout ("Show Rapicorn dialogs defined in GuiFile.xml.\n");
  printout ("This tool will read the GUI description file listed on the command\n");
  printout ("line, look for a dialog window and show it.\n");
  printout ("\n");
  printout ("Options:\n");
  printout ("  --parse-test                  Parse GuiFile.xml and exit.\n");
  printout ("  -x                            Enable auto-exit after first expose.\n");
  printout ("  --list                        List parsed definitions.\n");
  printout ("  --fatal-warnings              Turn criticals/warnings into fatal conditions.\n");
  printout ("  --snapshot pngname            Dump a snapshot to <pngname>.\n");
  printout ("  --test-dump                   Dump test stream after first expose.\n");
  printout ("  --test-matched-node PATTERN   Filter nodes in test dumps.\n");
  printout ("  -h, --help                    Display this help and exit.\n");
  printout ("  -v, --version                 Display version and exit.\n");
}

static bool parse_test = false;
static bool auto_exit = false;
static bool list_definitions = false;
static String dump_snapshot = "";
static bool test_dump = false;
static vector<String> test_dump_matched_nodes;

static void
parse_args (int    *argc_p,
            char ***argv_p)
{
  char **argv = *argv_p;
  uint argc = *argc_p;

  for (uint i = 1; i < argc; i++)
    {
      if (strcmp (argv[i], "--parse-test") == 0)
        {
          parse_test = true;
          argv[i] = NULL;
        }
      else if (strcmp (argv[i], "-x") == 0)
        {
          auto_exit = true;
          argv[i] = NULL;
        }
      else if (strcmp (argv[i], "--list") == 0)
        {
          list_definitions = true;
          argv[i] = NULL;
        }
      else  if (strcmp ("--snapshot", argv[i]) == 0 ||
                strncmp ("--snapshot=", argv[i], 11) == 0)
        {
          char *v = NULL, *equal = argv[i] + 10;
          if (*equal == '=')
            v = equal + 1;
          else if (i + 1 < argc)
            {
              argv[i++] = NULL;
              v = argv[i];
            }
          if (v)
            dump_snapshot = v;
          argv[i] = NULL;
        }
      else if (strcmp (argv[i], "--test-dump") == 0)
        {
          test_dump = true;
          argv[i] = NULL;
        }
      else  if (strcmp ("--test-matched-node", argv[i]) == 0 ||
                strncmp ("--test-matched-node=", argv[i], 20) == 0)
        {
          char *v = NULL, *equal = argv[i] + 19;
          if (*equal == '=')
            v = equal + 1;
          else if (i + 1 < argc)
            {
              argv[i++] = NULL;
              v = argv[i];
            }
          if (v)
            test_dump_matched_nodes.push_back (v);
          argv[i] = NULL;
        }
      else if (strcmp (argv[i], "--help") == 0 || strcmp (argv[i], "-h") == 0)
        {
          help_usage (false);
          exit (0);
        }
      else if (strcmp (argv[i], "--version") == 0 || strcmp (argv[i], "-v") == 0)
        {
          printout ("rapidrun (Rapicorn utilities) %s (Build ID: %s)\n", rapicorn_version().c_str(), rapicorn_buildid().c_str());
          printout ("Copyright (C) 2007 Tim Janik.\n");
          printout ("This is free software and comes with ABSOLUTELY NO WARRANTY; see\n");
          printout ("the source for copying conditions. Sources, examples and contact\n");
          printout ("information are available at https://rapicorn.testbit.org/.\n");
          exit (0);
        }
    }

  uint e = 1;
  for (uint i = 1; i < argc; i++)
    if (argv[i])
      {
        argv[e++] = argv[i];
        if (i >= e)
          argv[i] = NULL;
      }
  *argc_p = e;
}

static void
window_displayed (WindowH &window)
{
  if (!dump_snapshot.empty())
    {
      window.snapshot (dump_snapshot);
    }
  if (test_dump)
    {
#if 0 // FIXME
      for (uint i = 0; i < test_dump_matched_nodes.size(); i++)
        tstream->filter_matched_nodes (test_dump_matched_nodes[i]);
#endif
      String s = window.test_dump();
      printout ("%s", s.c_str());
      test_dump = false;
    }
  if (auto_exit)
    window.close();
}

static Blob
unit_test_res (const String &res_path)
{
  const int saved_errno = errno;
  const char *rres = getenv ("RAPIDRUN_RES");
  if (rres && strlen (rres) > 0)
    {
      if (Path::isabs (res_path))
        return Blob::load (res_path);
      StringVector dirs = string_split_any (rres, ":;");
      for (auto dir : dirs)
        {
          const String path = Path::join (dir, res_path);
          if (Path::check (path, "e"))
            return Blob::load (path);
        }
    }
  if (0)
    printerr ("rapidrun: missing resource: %s\n", res_path);
  errno = saved_errno;
  return Blob();
}

extern "C" int
main (int   argc,
      char *argv[])
{
  /* initialize Rapicorn and its backend (X11) */
  ApplicationH app = Rapicorn::init_app ("Rapidrun", &argc, argv); // setup Rapicorn.Application

  struct TestRes : Res { using Res::utest_hook; };
  TestRes::utest_hook (unit_test_res);

  parse_args (&argc, &argv);
  if (argc != 2)
    help_usage (true);

  /* find GUI definition file, relative to CWD */
  String filename = app.auto_path (argv[1], ".");

  /* load GUI definitions, fancy version of app.auto_load() */
  StringSeq definitions = app.auto_load (filename, "");

  /* print definitions */
  String dialog;
  for (uint i = 0; i < definitions.size(); i++)
    {
      bool iswindow = app.factory_window (definitions[i]);
      if (list_definitions)
        printout ("%s%s\n", definitions[i].c_str(), iswindow ? " (window)" : "");
      if (dialog == "" && iswindow)
        dialog = definitions[i];
    }
  if (list_definitions)
    exit (0);

  /* bail out without any dialogs */
  if (dialog == "")
    {
      printerr ("%s: no dialog definitions\n", filename.c_str());
      return 1;
    }

  // create window, hook up post-display handler
  WindowH window = app.create_window (dialog);
  window.sig_displayed() += [&window]() { window_displayed (window); };

  /* show window and process events */
  window.show();

  return app.run_and_exit();
}

} // anon
